Objects/odictobject.c

```diff
@@ -806,30 +807,50 @@ _odict_keys_equal(PyODictObject *a, PyODictObject *b)
 {
     _ODictNode *node_a, *node_b;
 
+    // keep operands' state and size to detect undesired mutations
+    const Py_ssize_t size_a = PyODict_SIZE(a);
+    const Py_ssize_t size_b = PyODict_SIZE(b);
+    if (size_a == -1 || size_b == -1) {
+        return -1;
+    }
+    if (size_a != size_b) {
+        // dictionaries with different number of keys can never be equal
+        return 0;
+    }
+
+    const size_t state_a = a->od_state;
+    const size_t state_b = b->od_state;
+
     node_a = _odict_FIRST(a);
     node_b = _odict_FIRST(b);
-    while (1) {
-        if (node_a == NULL && node_b == NULL)
-            /* success: hit the end of each at the same time */
-            return 1;
-        else if (node_a == NULL || node_b == NULL)
-            /* unequal length */
+    while (node_a != NULL && node_b != NULL) {
+        int res = PyObject_RichCompareBool((PyObject *)_odictnode_KEY(node_a),
+                                           (PyObject *)_odictnode_KEY(node_b),
+                                           Py_EQ);
```

Comment:
Bad things can happen if the key is disappear inside `PyObject_RichCompareBool()`. We should hold references to keys:
```c
PyObject *key_a = Py_NewRef(_odictnode_KEY(node_a));
...
Py_DECREF(key_a);
```

Reply:
Wouldn't it be impossible because of: https://github.com/python/cpython/issues/119004#issuecomment-2108428409?

Reply:
(Ah, maybe you mean as an additional step to use?)

Reply:
Yes, it is not enough, because nodes themself can be freed. But it may be needed for other issue, which is difficult to reproduce (or impossible in Python). So I have no reproducer, it is just my gut says that.